### PR TITLE
Feature/rectanglezoom

### PR DESCRIPTION
**Pull request by chris938626**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -81,9 +81,13 @@
...
     "tslint": "~5.8.0",
     "typescript": "~2.6.0",
-    "zone.js": "^0.8.4"
+    "zone.js": "^0.8.4",
+      "angular2-draggable": "^1.3.1"
   },
   "engines": {
     "node": ">=6.0.0"
+  },
+  "dependencies": {
```
Comment (hpawe01) [inline package.json:89]: Do we really need this as dependency? If yes, than we don't need it additionally as devDependency

File: src/index.ts [hunk 2]
```diff
@@ -8,6 +8,9 @@ import { NgxGalleryImageComponent } from './ngx-gallery-image.component';
 import { NgxGalleryThumbnailsComponent } from './ngx-gallery-thumbnails.component';
 import { NgxGalleryPreviewComponent } from './ngx-gallery-preview.component';
 import { NgxGalleryComponent } from './ngx-gallery.component';
+import { NgxGalleryScrollOverviewComponent } from "./ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component";
```
Comment (hpawe01) [inline src/index.ts]: Single quotes for imports please (everywhere)

File: src/ngx-gallery-action.component.ts [hunk 3]
```diff
@@ -1,23 +1,30 @@
-import { ChangeDetectionStrategy, Component, Input, Output, EventEmitter } from '@angular/core';
+import { ChangeDetectionStrategy, Component, Input, Output, EventEmitter, ViewChild, ElementRef } from '@angular/core';
+import { NgxGalleryPreviewComponent } from "./ngx-gallery-preview.component";
```
Comment (hpawe01) [inline src/ngx-gallery-action.component.ts]: For this and next import you can use `from './'`, because you exported them in `./index.ts` (=barrel-file)

File: src/ngx-gallery-action.component.ts [hunk 4]
```diff
@@ -1,23 +1,30 @@
...
+import { NgxGalleryScrollOverviewComponent } from "./ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component";
 
 @Component({
     selector: 'ngx-gallery-action',
     template: `
-        <div class="ngx-gallery-icon" [class.ngx-gallery-icon-disabled]="disabled"
+        <i class="ngx-gallery-icon {{ icon }}" [class.ngx-gallery-icon-disabled]="disabled"
```
Comment (hpawe01) [inline src/ngx-gallery-action.component.ts]: Is this change necessary? It seems like a major change that would produce merge-conflicts when pulling changes from the origin projects.

File: src/ngx-gallery-action.component.ts [hunk 5]
```diff
@@ -1,23 +1,30 @@
...
             aria-hidden="true"
             title="{{ titleText }}"
-            (click)="handleClick($event)">
-                <i class="ngx-gallery-icon-content {{ icon }}"></i>
-        </div>`,
+            (click)="handleClick($event)" >
+           <ngx-gallery-scroll-overview #scrollOverviewComponent [gallery]="gallery" *ngIf="rectangleScroll"> </ngx-gallery-scroll-overview>
```
Comment (hpawe01) [inline src/ngx-gallery-action.component.ts]: Do you need the viewchild reference (`#scrollOverviewComponent`)?

File: src/ngx-gallery-action.component.ts [hunk 6]
```diff
@@ -1,23 +1,30 @@
...
+        </i>`,
     changeDetection: ChangeDetectionStrategy.OnPush
 })
 export class NgxGalleryActionComponent {
+
```
Comment (hpawe01) [inline src/ngx-gallery-action.component.ts]: Please do not add new random new lines; can lead to merge-conflicts

File: src/ngx-gallery-action.component.ts [hunk 7]
```diff
@@ -1,23 +1,30 @@
...
     @Input() icon: string;
     @Input() disabled = false;
     @Input() titleText = '';
+    @Input() rectangleScroll: boolean;
```
Comment (hpawe01) [inline src/ngx-gallery-action.component.ts]: wouldn't something like `showScrollOverview` be more precise?
Or, if we change the name of the component to `zoomOverviewComponent` => `showZoomOverview`

File: src/ngx-gallery-action.component.ts [hunk 8]
```diff
@@ -1,23 +1,30 @@
...
+    @Input() gallery: NgxGalleryPreviewComponent;
 
     @Output() onClick: EventEmitter<Event> = new EventEmitter();
 
+    @ViewChild('scrollOverviewComponent') scrollOverviewComponent;
```
Comment (hpawe01) [inline src/ngx-gallery-action.component.ts]: Do we need this?

File: src/ngx-gallery-preview.component.ts [hunk 9]
```diff
@@ -8,7 +18,7 @@ import { NgxGalleryHelperService } from './ngx-gallery-helper.service';
     selector: 'ngx-gallery-preview',
     template: `
         <ngx-gallery-arrows (onPrevClick)="showPrev()" (onNextClick)="showNext()" [prevDisabled]="!canShowPrev()" [nextDisabled]="!canShowNext()" [arrowPrevIcon]="arrowPrevIcon" [arrowNextIcon]="arrowNextIcon"></ngx-gallery-arrows>
-        <div class="ngx-gallery-preview-top">
+        <div class="ngx-gallery-preview-top">TTTTEST
```
Comment (hpawe01) [inline src/ngx-gallery-preview.component.ts]: Not necessary anymore I guess ;)

File: src/ngx-gallery-preview.component.ts [hunk 10]
```diff
@@ -17,15 +27,16 @@ import { NgxGalleryHelperService } from './ngx-gallery-helper.service';
                 <ngx-gallery-action *ngIf="rotate" [icon]="rotateRightIcon" (onClick)="rotateRight()"></ngx-gallery-action>
                 <ngx-gallery-action *ngIf="fullscreen" [icon]="'ngx-gallery-fullscreen ' + fullscreenIcon" (onClick)="manageFullscreen()"></ngx-gallery-action>
                 <ngx-gallery-action [icon]="'ngx-gallery-close ' + closeIcon" (onClick)="close()"></ngx-gallery-action>
+                <ngx-gallery-action #galleryContainer [rectangleScroll]="rectangleScroll" [gallery]="this" ></ngx-gallery-action>
             </div>
-        </div>
+        </div>
```
Comment (hpawe01) [inline src/ngx-gallery-preview.component.ts]: Please do not add whitespaces at the end of a line

File: src/ngx-gallery-preview.component.ts [hunk 11]
```diff
@@ -17,15 +27,16 @@ import { NgxGalleryHelperService } from './ngx-gallery-helper.service';
                 <ngx-gallery-action *ngIf="rotate" [icon]="rotateRightIcon" (onClick)="rotateRight()"></ngx-gallery-action>
                 <ngx-gallery-action *ngIf="fullscreen" [icon]="'ngx-gallery-fullscreen ' + fullscreenIcon" (onClick)="manageFullscreen()"></ngx-gallery-action>
                 <ngx-gallery-action [icon]="'ngx-gallery-close ' + closeIcon" (onClick)="close()"></ngx-gallery-action>
+                <ngx-gallery-action #galleryContainer [rectangleScroll]="rectangleScroll" [gallery]="this" ></ngx-gallery-action>
```
Comment (hpawe01) [inline src/ngx-gallery-preview.component.ts]: Ohe whitespace to much

File: src/ngx-gallery-preview.component.ts [hunk 12]
```diff
@@ -411,6 +426,12 @@ export class NgxGalleryPreviewComponent implements OnChanges {
             return false;
         }
 
+        if ( img.naturalHeight > window.innerHeight ){
+            this.rectangleScroll = true;
+        } else {
+            console.log("rectangle scroll NOT active")
```
Comment (hpawe01) [inline src/ngx-gallery-preview.component.ts]: Not necessary anymore

File: src/ngx-gallery-preview.component.ts [hunk 13]
```diff
@@ -116,12 +129,12 @@ export class NgxGalleryPreviewComponent implements OnChanges {
     }
 
     open(index: number): void {
-
```
Comment (hpawe01) [inline src/ngx-gallery-preview.component.ts:119]: Please do not remove new lines in isolation

File: src/ngx-gallery-preview.component.ts [hunk 14]
```diff
@@ -116,12 +129,12 @@ export class NgxGalleryPreviewComponent implements OnChanges {
...
         this.onOpen.emit();
 
         this.index = index;
         this.isOpen = true;
         this.show(true);
+        this.showNext();
```
Comment (hpawe01) [inline src/ngx-gallery-preview.component.ts]: Why is this necessary? Seems a little bit odd, that we call `showNext`, when we just called `open`

File: src/ngx-gallery-preview.component.ts [hunk 15]
```diff
@@ -277,7 +292,7 @@ export class NgxGalleryPreviewComponent implements OnChanges {
     }
 
     canDragOnZoom() {
-        return this.fullSize || (this.zoom && this.zoomValue > 1);
+        return (this.fullSize && this.rectangleScroll)|| (this.zoom && this.zoomValue > 1);
```
Comment (hpawe01) [inline src/ngx-gallery-preview.component.ts]: missing whitespace

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss [hunk 16]
```diff
@@ -0,0 +1,13 @@
+.image-box {
+    background-size: contain;
+    background-image:
+        url("http://blog.doctoroz.com/wp-content/uploads/2013/07/AP_07_30_2013_53726226_L.jpg");
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss]: Here you should take the image that is used in the gallery. Maybe the smaller image that is also used in the gallery? (I know we don't use this at the moment, but as soon as we provide a smaller image for the ngx-gallery (see https://www.npmjs.com/package/ngx-gallery#ngxgalleryimage) this component should also use it)

To make it dynamic you have to set the css attribute `background-image` on the image box via `ngStyle` (see https://stackoverflow.com/a/34875479/5816097)

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html [hunk 17]
```diff
@@ -0,0 +1,8 @@
+<div id="previewContainer" class="drag-boundary image-box"
+     [ngClass]="{ 'top-b': !edge?.top, 'bottom-b': !edge?.bottom, 'left-b': !edge?.left, 'right-b': !edge?.right }"
+     #myBounds>
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html]: Better name: `previewContainer`

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html [hunk 18]
```diff
@@ -0,0 +1,8 @@
+<div id="previewContainer" class="drag-boundary image-box"
+     [ngClass]="{ 'top-b': !edge?.top, 'bottom-b': !edge?.bottom, 'left-b': !edge?.left, 'right-b': !edge?.right }"
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html]: Why do you need this? I cannot find this css-classes

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html [hunk 19]
```diff
@@ -0,0 +1,8 @@
...
+     #myBounds>
+    <div id="zoomContainer" ngDraggable class="drag-block image-detail" (started)="onStart($event)"
+         (stopped)="onStop($event)"
+         (edge)="checkEdge($event)" [bounds]="myBounds" [inBounds]="inBounds">
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html]: inBounds could be set to true here because we do not want to change this. So `inBounds="true"`

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html [hunk 20]
```diff
@@ -0,0 +1,8 @@
+<div id="previewContainer" class="drag-boundary image-box"
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html]: Use kebab-case also for ids (`preview-container`)
You won't need css classes here if you use the id instead of the class in the scss file (you don't use drag-boundary anyway)

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html [hunk 21]
```diff
@@ -0,0 +1,8 @@
...
+     [ngClass]="{ 'top-b': !edge?.top, 'bottom-b': !edge?.bottom, 'left-b': !edge?.left, 'right-b': !edge?.right }"
+     #myBounds>
+    <div id="zoomContainer" ngDraggable class="drag-block image-detail" (started)="onStart($event)"
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html]: id: `zoom-container`, classes not necessary

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss [hunk 22]
```diff
@@ -0,0 +1,13 @@
+.image-box {
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss]: use `#preview-container`

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss [hunk 23]
```diff
@@ -0,0 +1,13 @@
...
+    background-size: contain;
+    background-image:
+        url("http://blog.doctoroz.com/wp-content/uploads/2013/07/AP_07_30_2013_53726226_L.jpg");
+    position: absolute;
+    top: 50vh;
+    left:-300px;
+}
+
+.image-detail {
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss]: use `#zoom-container`

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss [hunk 24]
```diff
@@ -0,0 +1,13 @@
...
+    opacity: 0.5;
+    padding:0;
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss]: missing whitespace

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss [hunk 25]
```diff
@@ -0,0 +1,13 @@
+.image-box {
+    background-size: contain;
+    background-image:
+        url("http://blog.doctoroz.com/wp-content/uploads/2013/07/AP_07_30_2013_53726226_L.jpg");
+    position: absolute;
+    top: 50vh;
+    left:-300px;
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss]: missing whitespace

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+export class NgxGalleryScrollOverviewComponent implements AfterViewInit{
+    private beforeZoomLeft: number;
+    private beforeZoomTop: number;
+    private initialLeft: number;
+    private initialTop: number;
+    private SCALE_FACTOR = 8;
+
+    @Input() gallery: NgxGalleryPreviewComponent;
+
+    inBounds = true;
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: not necessary

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+        this.updatePreviewScales();
+    }
+
+    checkEdge(event) {
+        this.edge = event;
+    }
+
+    onStop(event) {
+        var rect = event.getBoundingClientRect();
+        this.gallery.positionLeft += (this.beforeZoomLeft - rect.left) * this.SCALE_FACTOR;
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: Maybe it is possible to realize this functionality via events. That way you don't have to pass through the whole gallery.

Event would be something like `zoomContainerChanged` and in the gallery-component you can listen to this event and handle the position accordingly

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+        document.getElementById('previewContainer').style.height = heightScaled + 'px';
+
+        // scale the zoom area for visualization which area the user sees
+        var zoomHeightScaled = window.innerHeight / this.SCALE_FACTOR;
+        var zoomWidthScaled = window.innerWidth / this.SCALE_FACTOR;
+        document.getElementById('zoomContainer').style.width = zoomWidthScaled + 'px';
+        document.getElementById('zoomContainer').style.height = zoomHeightScaled + 'px';
+
+        // init
+        this.initialLeft = this.gallery.positionLeft;
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: Analog to the event (see below) you could pass down the position and use it here

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+        document.getElementById('zoomContainer').style.transform = transformString;
+    }
+
+    public updatePreviewScales() {
+        var img = document.getElementsByClassName('ngx-gallery-fullsize');
+        var width = img[0].clientWidth;
+        var height = img[0].clientHeight;
+
+        // set the background image of the preview Container which contains the Scaled preview Image
+        document.getElementById('previewContainer').style.backgroundImage = "url('"+ img[0].getAttribute('src') +"')";
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: see my comment in the css file

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 30]
```diff
@@ -0,0 +1,81 @@
...
+
+        // scale the preview image
+        var widthScaled = width / this.SCALE_FACTOR;
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: Define those variables at the beginning of the component and use it in the template via ngStyle

**Review by hpawe01 — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 31]
```diff
@@ -8,6 +8,9 @@ import { NgxGalleryImageComponent } from './ngx-gallery-image.component';
...
+import { AngularDraggableModule } from "angular2-draggable";
+
```
Comment (hpawe01) [inline src/index.ts]: No additional new lines

File: package.json (hunk 1 above)
Comment (chris938626) [reply package.json:89]: done

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts (hunk 27 above)
Comment (chris938626) [reply src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: done

File: src/index.ts (hunk 2 above)
Comment (chris938626) [reply src/index.ts]: done

File: src/ngx-gallery-action.component.ts (hunk 3 above)
Comment (chris938626) [reply src/ngx-gallery-action.component.ts]: done

File: src/ngx-gallery-action.component.ts (hunk 4 above)
Comment (chris938626) [reply src/ngx-gallery-action.component.ts]: done

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html [hunk 32]
```diff
@@ -1,8 +1,7 @@
-<div id="previewContainer" class="drag-boundary image-box"
-     [ngClass]="{ 'top-b': !edge?.top, 'bottom-b': !edge?.bottom, 'left-b': !edge?.left, 'right-b': !edge?.right }"
+<div id="preview-container" [ngStyle]="previewContainerStyles"
      #myBounds>
-    <div id="zoomContainer" ngDraggable class="drag-block image-detail" (started)="onStart($event)"
+    <div id="zoom-container" [ngStyle]="zoomContainerStyles" ngDraggable (started)="onStart($event)"
          (stopped)="onStop($event)"
-         (edge)="checkEdge($event)" [bounds]="myBounds" [inBounds]="inBounds">
+         (edge)="checkEdge($event)" [bounds]="myBounds" [inBounds]="true">
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.html]: Do we still need the `edge` event handler?

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss [hunk 33]
```diff
@@ -1,11 +1,14 @@
...
+    overflow: hidden;
 }
 
-.image-detail {
+#zoom-container {
     opacity: 0.5;
-    padding:0;
+    padding: 0;
+    background: black;
+    cursor: col-resize;
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.scss]: `move` is better

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 34]
```diff
@@ -1,65 +1,62 @@
...
     private initialLeft: number;
     private initialTop: number;
     private SCALE_FACTOR = 8;
 
-    @Input() gallery: NgxGalleryPreviewComponent;
+    @Output() onZoomChanged = new EventEmitter();
+    @Input() zoomPosition: ZoomPosition;
 
-    inBounds = true;
     edge = {
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: Do we still need this?

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 35]
```diff
@@ -1,65 +1,62 @@
...
     }
 
     public updateDetailZoom() {
-        var topScaled = (this.initialTop - this.gallery.positionTop) / this.SCALE_FACTOR;
-        var leftScaled = (this.initialLeft - this.gallery.positionLeft) / this.SCALE_FACTOR;
+        var topScaled = (this.initialTop - this.zoomPosition.positionTop) / this.SCALE_FACTOR;
+        var leftScaled = (this.initialLeft - this.zoomPosition.positionLeft) / this.SCALE_FACTOR;
         var transformString = 'translate(' + leftScaled + 'px,' + topScaled + 'px)';
-        document.getElementById('zoomContainer').style.transform = transformString;
+        this.zoomContainerStyles["transform"] = transformString;
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: It is considered best practice to use dot-notation when possible. So `this.zoomContainerStyles.transform` would be better. Also: you don't really need to define the `transformString` one line before if you just assign it here.

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 36]
```diff
@@ -1,65 +1,62 @@
...
 
-        // set the background image of the preview Container which contains the Scaled preview Image
-        document.getElementById('previewContainer').style.backgroundImage = "url('" + img[0].getAttribute('src') + "')";
-
         // scale the preview image
-        var widthScaled = width / this.SCALE_FACTOR;
-        var heightScaled = height / this.SCALE_FACTOR;
-        document.getElementById('previewContainer').style.width = widthScaled + 'px';
-        document.getElementById('previewContainer').style.height = heightScaled + 'px';
+        this.previewContainerStyles["background-image"] = "url('" + img[0].getAttribute('src') + "')";
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: Consider using single quotes in javascript and double quotes in css and html: `this.previewContainerStyles['background-image'] = 'url("' + img[0].getAttribute('src') + '")';`

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 37]
```diff
@@ -1,65 +1,62 @@
...
+        this.previewContainerStyles["height"] = (height / this.SCALE_FACTOR) + 'px';
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: Use dot-notation for next four lines (e.g. `this.previewContainerStyles.height`)

File: src/ngx-gallery-action.component.ts [hunk 38]
```diff
@@ -1,30 +1,35 @@
...
     selector: 'ngx-gallery-action',
     template: `
-        <i class="ngx-gallery-icon {{ icon }}" [class.ngx-gallery-icon-disabled]="disabled"
+        <div class="ngx-gallery-icon {{ icon }}" [class.ngx-gallery-icon-disabled]="disabled"
             aria-hidden="true"
             title="{{ titleText }}"
             (click)="handleClick($event)" >
-           <ngx-gallery-scroll-overview #scrollOverviewComponent [gallery]="gallery" *ngIf="rectangleScroll"> </ngx-gallery-scroll-overview>
-        </i>`,
+           <ngx-gallery-scroll-overview #scrollOverviewComponent (onZoomChanged)="zoomChanged()" [zoomPosition] = "zoomPosition" *ngIf="showScrollOverview"> </ngx-gallery-scroll-overview>
```
Comment (hpawe01) [inline src/ngx-gallery-action.component.ts]: Yeah, this is what I meant. To make event more use of EventEmitter you should pass in the event (and the event could be an object like `{before: ZoomPosition after: ZoomPosition}`, but this would be defined in the `ScrollOverviewComponent`). And switch the names: The event should be named `zoomChanged`, because that is what just happend and the responding function should be named `onZoomChanged`, because this is the reaction to the event. Result would be: `(zoomChanged)="onZoomChanged($event)"`

File: src/ngx-gallery-action.component.ts [hunk 39]
```diff
@@ -1,30 +1,35 @@
...
     @Input() icon: string;
     @Input() disabled = false;
     @Input() titleText = '';
-    @Input() rectangleScroll: boolean;
+    @Input() showScrollOverview: boolean;
+    @Input() zoomPosition: ZoomPosition;
     @Input() gallery: NgxGalleryPreviewComponent;
 
     @Output() onClick: EventEmitter<Event> = new EventEmitter();
+    @Output() onZoomChanged = new EventEmitter();
```
Comment (hpawe01) [inline src/ngx-gallery-action.component.ts]: Change name to `zoomChanged` (I know, they use `onClick` for their event emitter but this just makes no sense.

File: src/ngx-gallery-action.component.ts [hunk 40]
```diff
@@ -1,30 +1,35 @@
...
 
     @ViewChild('scrollOverviewComponent') scrollOverviewComponent;
 
+    zoomChanged() {
+        this.onZoomChanged.emit();
```
Comment (hpawe01) [inline src/ngx-gallery-action.component.ts]: Here you would pass through the `event`

File: src/ngx-gallery-preview.component.ts [hunk 41]
```diff
@@ -27,16 +28,16 @@ import { NgxGalleryHelperService } from './ngx-gallery-helper.service';
                 <ngx-gallery-action *ngIf="rotate" [icon]="rotateRightIcon" (onClick)="rotateRight()"></ngx-gallery-action>
                 <ngx-gallery-action *ngIf="fullscreen" [icon]="'ngx-gallery-fullscreen ' + fullscreenIcon" (onClick)="manageFullscreen()"></ngx-gallery-action>
                 <ngx-gallery-action [icon]="'ngx-gallery-close ' + closeIcon" (onClick)="close()"></ngx-gallery-action>
-                <ngx-gallery-action #galleryContainer [rectangleScroll]="rectangleScroll" [gallery]="this" ></ngx-gallery-action>
+                <ngx-gallery-action *ngIf="zoomPosition" #galleryContainer (onZoomChanged) = "zoomChanged()" [zoomPosition] = "zoomPosition" [showScrollOverview]="showScrollOverview" [gallery]="this"></ngx-gallery-action>
```
Comment (hpawe01) [inline src/ngx-gallery-preview.component.ts]: See remarks regarding naming and `$event` for `zoomChanged` above

File: src/ngx-gallery-preview.component.ts [hunk 42]
```diff
@@ -113,6 +115,11 @@ export class NgxGalleryPreviewComponent implements OnChanges {
         }
     }
 
+    zoomChanged() {
```
Comment (hpawe01) [inline src/ngx-gallery-preview.component.ts:118]: Now here you would get the `event` and could access e.g. the current `ZoomPosition` with `event.after` (or you name the variable `zoomPositions`, so that the code reads more clearly => `zoomPositions.after`).

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 43]
```diff
@@ -1,65 +1,62 @@
...
-export class NgxGalleryScrollOverviewComponent implements AfterViewInit {
+export class NgxGalleryScrollOverviewComponent implements OnInit {
     private beforeZoomLeft: number;
     private beforeZoomTop: number;
     private initialLeft: number;
     private initialTop: number;
     private SCALE_FACTOR = 8;
 
-    @Input() gallery: NgxGalleryPreviewComponent;
+    @Output() onZoomChanged = new EventEmitter();
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: See naming remarks above

File: src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts [hunk 44]
```diff
@@ -69,8 +66,9 @@ export class NgxGalleryScrollOverviewComponent implements AfterViewInit {
 
     onStop(event) {
         var rect = event.getBoundingClientRect();
-        this.gallery.positionLeft += (this.beforeZoomLeft - rect.left) * this.SCALE_FACTOR;
-        this.gallery.positionTop += (this.beforeZoomTop - rect.top) * this.SCALE_FACTOR;
+        this.zoomPosition.positionLeft += (this.beforeZoomLeft - rect.left) * this.SCALE_FACTOR;
+        this.zoomPosition.positionTop += (this.beforeZoomTop - rect.top) * this.SCALE_FACTOR;
+        this.onZoomChanged.emit();
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/ngx-gallery-scroll-overview.component.ts]: And here you would emit the before and after zoom position: `this.zoomChanged.emit(zoomPositions)` (you have to build this object before of course).

File: src/ngx-gallery-scroll-overview/zoomPosition.ts [hunk 45]
```diff
@@ -0,0 +1,4 @@
+export class ZoomPosition {
```
Comment (hpawe01) [inline src/ngx-gallery-scroll-overview/zoomPosition.ts:1]: This is more an interface than a class: `export interface ZoomPosition`. And please change the filename to `zoom-position.interface.ts`

**Review by hpawe01 — CHANGES_REQUESTED** [review]
With the event emitter it makes much more sense. But we should use it to pass through the event wich contains the changes.
Other things are just minor format remarks.
